soc/usbexecd.c

```diff
@@ -0,0 +1,1029 @@
+        pb->bufcount -= written;
+
+        // If the start position marker is at the buffer end
+        if (pb->startpos == PIPE_BUF) {
+            // Set it at the start of the buffer for the next read
+            pb->startpos = 0;
+        }
+
+        write_len -= written;
+    }
+
+    return (total_written - write_len);
+}
+
+void modify_pipebuf_epoll(pipebuf_t* pb, int operation) {
+    // Create an epoll_event related to that file descriptor
+    struct epoll_event evt;
+    evt.data.ptr = pb;
+    evt.events = pb->events;
+
+    // Register the event
+    int r = epoll_ctl(epfd, operation, pb->fd, &evt);
+    // Verify it worked
+    if (r < 0) {
+        fatal("epoll_ctl failed: %s", strerror(errno));
+    }
+}
+
+void add_pipebuf_epoll(pipebuf_t* pb) {
+    pipebuf_common_debug(pb, "Adding pipe buf to epoll");
+    modify_pipebuf_epoll(pb, EPOLL_CTL_ADD);
+}
+
+void delete_pipebuf_epoll(pipebuf_t* pb) {
+    pipebuf_common_debug(pb, "DELETING pipe buf from epoll");
+    modify_pipebuf_epoll(pb, EPOLL_CTL_DEL);
+}
+
+// Closing functionality for read and write pipe buffers
+void pipebuf_common_close(pipebuf_t* pb) {
+    pipebuf_common_debug(pb, "Attempting to close");
+    // Mark this buffer as ready to close
+    pb->eof = true;
+
+    // If it still has data internally, don't close it
+    // If it does, go wild
+    if (pb->bufcount == 0) {
+        // If the file descriptor exists (which it should)
+        if (pb->fd != -1) {
+            // Close it
+            int ret = close(pb->fd);
+            // info("We attempted to close the file descriptor %d", ret);
+            if (ret == -1) {
+                fatal("Unable to close the file descriptor: %s", strerror(errno));
+            }
+            // Set it back to -1
+            pb->fd = -1;
+            // Reset other fields
+            pb->startpos = pb->endpos = pb->bufcount = 0;
+            pipebuf_common_debug(pb, "Successfully closed");
+        }
+    }
+    else {
+        pipebuf_common_debug(pb, "Requested Close but it has remaining bytes...");
+    }
+}
+
+void pipebuf_common_debug(pipebuf_t *pb, const char *str) {
+    debug("From process %d, role %d, credit %d, bufcount %d, startpos %d, endpos %d", pb->id, pb->role, pb->credit, pb->bufcount, pb->startpos, pb->endpos);
+    debug("\t%s", str);
+}
+
+int pipebuf_in_init(pipebuf_t* pb, int id, int role) {
+    int fd = pipebuf_common_init(pb, id, role, EPOLLIN, 0);
+
+    return fd;
+}
+
+void pipebuf_in_close(pipebuf_t* pb) {
+
+    // Close the pipebuf
+    pipebuf_common_close(pb);
+
+    // If this pipebuf was closed successfully
+    // (it could have bytes remaining in the pipe buffer)
+    if (pb->fd == -1) {
+        // Tell the other end of the pipe that this fd has been closed
+        send_header(CMD_CLOSE_CONTROL + pb->role, pb->id, 0, 0);
+    }
+}
+
+
+// INTERNAL -> CLI
+int pipebuf_in_write_to_sock(pipebuf_t* pb, size_t num_to_write) {
+
```

Comment:
Is there a reason not to just use `pb->bufcount` instead of accepting `num_to_write` as an argument?